tests/EntryPoints/Specials/SpecialManageApproversTest.php

```diff
@@ -64,18 +67,13 @@ public function testAddApproverAction(): void {
 			]
 		);
 
-		$output = $this->viewPage();
+		$approvers = $this->approverRepository->getApproversWithCategories();
 
-		$this->assertStringContainsString(
-			$username,
-			$output,
-			'Expected HTML output to contain the new approver username'
-		);
-		$this->assertStringNotContainsString(
-			'<div class="category-entry">',
-			$output,
-			'New approver should have no categories'
-		);
+		$approver = array_filter( $approvers, fn( $approver ) => $approver['userId'] === $userId );
+		$newApprover = reset( $approver );
+
+		$this->assertNotNull( $newApprover, 'New approver should be added' );
+		$this->assertSame( [], $newApprover['categories'], 'New approver should have no categories' );
```

Comment:
This drew my attention since logic such as the array_filter is a yellow flag in tests.

It seems OK in this case, since we do not have a proper ApproverList collection class that could house this logic. However it's not nice to pollute the test case with these lower level details. You need to read all of them to guess at the intent. Better to extract a well named function. Perhaps `assertNewApproverExists( int $userId ): void`, `getApproverWithUserId( int $userId ): ?Approver`, or some combination.